### PR TITLE
ENH: n-D and nan_policy support for scipy.stats.percentileofscore

### LINKED ISSUE #2868: nan and stats.percentileofscore
In function `stats.percentileofscore`, `nan` is (effectively) treated as +infinity.
E.g.

```
scipy.stats.percentileofscore ([3, 1, numpy.nan, numpy.nan, 2], 5)
```

returns `60.00` (as it the percent of values below 5).

However, other numerical functions (e.g. `scipy.mean`) return `nan` for entries containing `nan`s. 

Is it the desired behaviour of `stats.percentileofscore`?

### PR DESCRIPTION
**Pull request by patnr**
ENH: Implement nd support for percentileofscore

The enhancement is inspired by https://stackoverflow.com/a/12414469,
which (as necessary before this) uses pure python loops, which is slow.

Note that BrenBarn's top-ranked answer on that question uses `rankdata`,
which has been added to the 'see also' section. The method of `rankdata`
uses `argsort` (as in an older version of `percentileofscore`) and so
it is faster. For ex., for assigning percentiles to each value in an
array of length n `rankdata` is `O(n log n)`,
while this `percentileofscore` is `O(n^2)`.
However, the algorithm of `rankdata` is not equivalent when
the list contains duplicates.

The enchantment also required an overhaul of the treatment of nan's.

Closes #2868

### CODE REVIEW

**Comment by patricknraanes** [comment]
Here are some benchmarks to demonstrate the speed-up compared to using pure/python loops. It also compares to argsort/rankdata algorithms. It is based on the answer from BrenBarn in the stackoverflow question mentioned above.

    # Speed test.
    # NB: Note that the argsort and rankdata methods
    # are not equivalent to percentileofscore
    # when there are ties/duplicates.

    import numpy as np
    import scipy.stats as stats

    def list_to_percentiles(numbers):
        pairs = list(zip(numbers, range(len(numbers))))
        pairs.sort(key=lambda p: p[0])
        result = [0 for i in range(len(numbers))]
        for rank in range(len(numbers)):
            original_index = pairs[rank][1]
            result[original_index] = rank * 100.0 / (len(numbers)-1)
        return result

    print("\nSMALL DATA:")
    x = [1, 1, 2, 2, 17]

    print("Python loop around percentileofscore:")
    %timeit [stats.percentileofscore(x, i) for i in x]
    # 223 µs ± 6.7 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

    print("new nd percentileofscore:")
    %timeit stats.percentileofscore(x, x)
    # 47.3 µs ± 1.63 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

    print("Argsort method:")
    %timeit list_to_percentiles(x)
    # 2.47 µs ± 45.2 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

    print("rankdata:")
    %timeit stats.rankdata(x, "average")/len(x)
    # 44.9 µs ± 337 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

    print("\nLARGE DATA:")
    x = np.random.randint(0, 10000, 1000)

    print("Python loop around percentileofscore:")
    %timeit [stats.percentileofscore(x, i) for i in x]
    # 50 ms ± 747 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

    print("new nd percentileofscore:")
    %timeit stats.percentileofscore(x, x)
    # 3.39 ms ± 58.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

    print("Argsort method:")
    %timeit list_to_percentiles(x)
    # 541 µs ± 9.11 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

    print("rankdata:")
    %timeit stats.rankdata(x, "average")/len(x)
    # 103 µs ± 2.21 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

**Comment by patricknraanes** [comment]
BTW, https://github.com/scipy/scipy/issues/2868 can be closed (also before this PR) as `percentileofscore` no longer uses sorting anyhow, and nan's are properly treated (propagated).

**Comment by mdhaber** [comment]
In the master branch, `scipy.stats.percentileofscore ([3, 1, numpy.nan, numpy.nan, 2], 5)` still returns 60 on my machine.

**Comment by patricknraanes** [comment]
> In the master branch, `scipy.stats.percentileofscore ([3, 1, numpy.nan, numpy.nan, 2], 5)` still returns 60 on my machine. It doesn't on yours?

My bad. I thought the issue was with nan's in the second arg, which worked fine in this PR.

Seeing as I'm revising the function as a whole anyway, I think I should be able to address #2868 as well. Will give it a try.

**Comment by patricknraanes** [comment]
I believe the commit now fixes #2868 as well.

Notes about the implementation:
- In order to support lists containing both + and - infinity at the same time, I needed to re-implement `_contains_nan` to use a more basic/slow test.
- Comparing to an empty list now produces nan, not 100. This makes a whole lot more sense to me. Consider: why would the percentile of a score vs an empty list be 100, and not 0.

**Comment by patricknraanes** [comment]
Ping @mdhaber 

BTW, do you know/think I should post on the scipy mailinglist for this PR?

**Comment by mdhaber** [comment]
That might help attract a reviewer. I'd suggest addressing the failing CI first, though.

**Comment by tupui** [comment]
It's a quality PR! Could you add tests for the new policy?

**Comment by patricknraanes** [comment]
Thanks!

I've added more tests, and implemented a kwarg switch to `_contains_nan` that enables using the safer version (that does not fail with arrays containing both +/- infinity).

**Review by tupui — CHANGES_REQUESTED** [review]
Thanks for updating the PR with all these tests! I would just propose to adjust a few things.

File: scipy/stats/tests/test_stats.py [hunk 1]
```diff
@@ -2829,6 +2829,32 @@ def test_percentileofscore():
 
     assert_raises(ValueError, pcos, [1, 2, 3, 3, 4], 3, kind='unrecognized')
 
+    # score *arrays*
+    a = 1 + np.arange(10)
+    assert_equal(pcos(a, [4, 5]), [40.0, 50.0])
```
[the remark this answers is not in the document]
Comment (patricknraanes) [reply scipy/stats/tests/test_stats.py:2834]: Thanks for the input!

I followed the example of the other tests for percentileofscore.

I believe if I am to use `@pytest.mark.parametrize` then I would have to move the tests out of `test_percentileofscore()`, which seems not so nice. Am I understanding you correctly?

File: scipy/stats/tests/test_stats.py [hunk 2]
```diff
@@ -2829,6 +2829,32 @@ def test_percentileofscore():
...
+    assert_equal(pcos([1,np.inf], [-np.inf, +np.inf], kind="mean"),   [0.0, 75.0])
+    assert_equal(pcos([1,np.inf], [-np.inf, +np.inf], kind="weak"),   [0.0, 100.0])
+
+    # nan's
+    assert_equal(pcos([], 1), np.nan)
+    assert_equal(pcos([np.nan, 1, 2],   [0, 1, 2], nan_policy='omit'),      [0, 50.0, 100.0])
+    assert_equal(pcos([np.nan, 2],      [0, 1, 2], nan_policy='propagate'), [np.nan, np.nan, np.nan])
+    assert_equal(pcos([np.nan, np.nan], [0, 1, 2], nan_policy='omit'),      [np.nan, np.nan, np.nan])
+    assert_equal(pcos([1, 2], [1, 2, np.nan], nan_policy='propagate'),      [50.0, 100.0, np.nan])
+    assert_raises(ValueError, pcos, [1, 2, 3, np.nan], [1, 2, 3], nan_policy='raise')
```
[the remark this answers is not in the document]
Comment (patricknraanes) [reply scipy/stats/tests/test_stats.py]: I believe this is resolved now

[the remark this answers is not in the document]
Comment (tupui) [reply scipy/stats/tests/test_stats.py]: It is thank you! Feel free to improve existing tests as well (not required of course).

File: scipy/stats/tests/test_stats.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (patricknraanes) [reply scipy/stats/tests/test_stats.py]: Ok, I've refactored the tests to use a class and `@pytest.mark.parametrize`. Let me know what you think!

[the remark this answers is not in the document]
Comment (tupui) [reply scipy/stats/tests/test_stats.py]: It's better thanks! Also some cases were not covered which you fixed.

I compared both versions and if I am not mistaking it matches. I think I am just missing this one (I think the one called `test_large_numbers_missing` should have been it):

```python
    for kind in ('rank', 'mean', 'strict', 'weak'):
        assert_equal(
              pcos([10, 20, 30, 50, 60, 70, 80, 90, 100, 110],
                   40, kind=kind), 30.0)
```

A last thing, I know it is not good for presentation, but we should not align list elements. So make sure to have just a single space after a comma. Not less, not more.

[the remark this answers is not in the document]
Comment (patricknraanes) [reply scipy/stats/tests/test_stats.py]: You're right, I had omitted that one (and wrongly named another one).

Also fixed the spacing, as requested.

File: scipy/stats/tests/test_stats.py [hunk 3]
```diff
@@ -2758,80 +2758,84 @@ class TestsPercentileOfScore:
...
-    @pytest.mark.parametrize("kind, result", [("rank",   40),
-                                              ("mean",   35),
+    @pytest.mark.parametrize("kind, result", [("rank", 40),
+                                              ("mean", 35),
                                               ("strict", 30),
-                                              ("weak",   40)])
+                                              ("weak", 40)])
     def test_unique(self, kind, result):
-        assert_equal(self.f([1,2,3,4,5,6,7,8,9,10], 4, kind=kind), result)
+        assert_equal(self.f([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 4, kind=kind),  result)
```
Comment (tupui) [inline scipy/stats/tests/test_stats.py:2766]: Great thanks! Just this one is extra.
```suggestion
        assert_equal(self.f([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 4, kind=kind), result)
```

File: scipy/stats/tests/test_stats.py [hunk 4]
```diff
@@ -2752,82 +2752,95 @@ def test_1samp_alternative(self):
...
+                              ("omit", [np.nan, np.nan], [0, 1, 2], [np.nan, np.nan, np.nan]),
+                              ])
+    def test_nans_ok(self, policy, a, score, result):
+        assert_equal(self.f(a, score, nan_policy=policy), result)
+
+    @pytest.mark.parametrize("policy, a, score, message", [
+        ("raise", [1, 2, 3, np.nan], [1, 2, 3], "The input contains nan values"),
+        ("raise", [1, 2, 3], [1, 2, 3, np.nan], "The input contains nan values"),
+        ("omit", [1, 2, 3], [1, 2, 3, np.nan], ("The input scores contains nan values,"
+                                                 " which is incompatible with the 'omit' policy.")),
```
Comment (tupui) [inline scipy/stats/tests/test_stats.py:2839]:
```suggestion
        ("omit", [1, 2, 3], [1, 2, 3, np.nan], ("The input scores contains nan values,"
                                                " which is incompatible with the 'omit' policy.")),
```

**Comment by patricknraanes** [comment]
So some Azure pipeline tests are failing. I don't know why it's them in particular. The relevant part of the error messages is this:

```
scipy/stats/stats.py:2182: in percentileofscore
    left  = count(a < score)
E   RuntimeWarning: invalid value encountered in less
        a          = masked_array(data=[1.0, 2.0, --],
                                  mask=[False, False,  True],
                                  fill_value=1e+20)
        cna        = True
        cns        = False
        count      = <function percentileofscore.<locals>.<lambda> at 0x7f380cca2560>
        kind       = 'rank'
        n          = 2
        nan_policy = 'omit'
        npa        = 'omit'
        nps        = 'omit'
        score      = array([[0],
                            [1],
                            [2]])
```

So it seems the masked arrays don't work to hide nan's on this platform. Is this a known issue?

File: scipy/stats/stats.py [hunk 5]
```diff
@@ -2104,28 +2116,90 @@ def percentileofscore(a, score, kind='rank'):
...
+            right = count(a <= score)
+            plus1 = left < right
+            perct = (left + right + plus1) * (50.0 / n)
+        elif kind == 'strict':
+            perct = count(a < score) * (100.0 / n)
+        elif kind == 'weak':
+            perct = count(a <= score) * (100.0 / n)
+        elif kind == 'mean':
+            left  = count(a < score)
+            right = count(a <= score)
```
Comment (tupui) [inline scipy/stats/stats.py:2192]: Seems like you would need to use `numpy.ma.masked_less` instead!

Comment (patricknraanes) [reply scipy/stats/stats.py:2192]: Thanks! That's interesting. Indeed, I believe it could be worked using `np.ma.masked_less` & co. However, it seems like there would be an inefficiency in that: these routines create a new `MaskedArray`, whereas `a < score` simply creates the boolean mask.

Furthermore, although I'm not well versed with masked arrays, the error message produced on Azure pipeline (presumably complaining about nan's) makes little sense to me. Operations on `MaskedArray`s are not really supposed to process the masked elements. Furthermore, the error does not seem to occur on Linux/Mac platforms, adding to my suspicion that something funny is going on. I note that the error is a `RunTimeWarning`, so maybe it could just be silenced?

Comment (tupui) [reply scipy/stats/stats.py:2192]: I never worked with masked array. Maybe ask another maintainer here.

But sure, you could try to silent the warning and see if tests pass.

Comment (patricknraanes) [reply scipy/stats/stats.py]: Seems like that did it 🥳

Comment (tupui) [reply scipy/stats/stats.py]: GG. We just need to have another eye on this as I have no experience with masked arrays.

**Review by tupui — APPROVED** [review]
LGTM. @mdhaber I am not sure about masked array, if you could have a second look that would be great. Thanks.

**Comment by mdhaber** [comment]
OK, I'll try to take a look this weekend.

**Comment by mdhaber** [comment]
Is this PR still intended to add n-D array support? If so, there would need to be a new `axis` argument. 

The example says:
```
    Multidimensional score arrays are supported:
    >>> stats.percentileofscore([1, 2, 3, 3, 4], [2, 3])
    array([40., 70.])
```
but these are still just two 1D arrays.

The thing that is new is that the second argument can be a 1D array instead of just a scalar, right? If so, please adjust the title and description of the example accordingly.

File: scipy/stats/tests/test_stats.py [hunk 6]
```diff
@@ -2753,82 +2753,94 @@ def test_1samp_alternative(self):
...
-                                  ('mean', 0, 0.0)]:
-        assert_equal(
-              pcos([10, 20, 30, 50, 60, 70, 80, 90, 100, 110],
-                   score, kind=kind), result)
-
-    assert_raises(ValueError, pcos, [1, 2, 3, 3, 4], 3, kind='unrecognized')
+class TestPercentileOfScore:
+
+    def f(self, *args, **kwargs):
+        return stats.percentileofscore(*args, **kwargs)
```
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:2760]: Is this just to save horizontal space? e,g,
`self.f`
vs
`stats.percentileofscore`?

File: scipy/stats/stats.py (hunk 5 above)
Comment (mdhaber) [reply scipy/stats/stats.py]: @patricknraanes did you experiment at all with masking the arrays yourself (e.g. `x = x[~np.isnan(x)]`) rather than using masked arrays?

File: scipy/stats/tests/test_stats.py (hunk 6 above)
Comment (patricknraanes) [reply scipy/stats/tests/test_stats.py:2760]: Yes. No good?

**Comment by patricknraanes** [comment]
> The thing that is new is that the second argument can be a 1D array instead of just a scalar, right? If so, please adjust the title and description of the example accordingly.

The second argument (`score`) can be of any dimension. However, I expect that the most common use case of this PR would be to score an array against itself, i.e. two 1d arrays, so I didn't add examples with higher-dim `score`. Still, I guess I should add a test for it.

I'm not sure if an `axis` argument makes sense here.  The operation is by definition per-element (of `score`). I suppose if the first argument (`a`) were allowed to be 2d or higher-dim, then it might make sense. I could look into supporting that, although I feel like it is of much less interest than the current extension.

File: scipy/stats/stats.py (hunk 5 above)
Comment (patricknraanes) [reply scipy/stats/stats.py]: As I recall, I think I did. I don't remember very well, but I think I chose the `masked_array` approach because it made it easier to support all cases of `nan_policy`.

File: scipy/stats/tests/test_stats.py (hunk 6 above)
Comment (mdhaber) [reply scipy/stats/tests/test_stats.py:2760]: No judgement yet : ) Yup, I saw that it was like that. Just checking.

**Comment by mdhaber** [comment]
A test for n-d second argument would be good.
As written, the PR doesn't need an `axis` argument. Only needed if the first argument we're to support n-d arrays. 
No need to change right now. I'm just thinking ahead and making sure we don't do something here that would make it harder to extend later.

File: scipy/stats/stats.py (hunk 5 above)
Comment (mdhaber) [reply scipy/stats/stats.py]: Ok. Probably fine. I haven't looked closely about which of the changes are related to masked array and which are due to array second argument. If I see major simplifications that would result from manual masking, i may suggest them. (Masked arrays tend to be slower in my experience, so they would have to make the code simpler to justify them.)

File: scipy/stats/stats.py [hunk 7]
```diff
@@ -223,16 +224,20 @@
...
     if nan_policy not in policies:
         raise ValueError("nan_policy must be one of {%s}" %
                          ', '.join("'%s'" % s for s in policies))
     try:
-        # Calling np.sum to avoid creating a huge array into memory
-        # e.g. np.isnan(a).any()
-        with np.errstate(invalid='ignore'):
-            contains_nan = np.isnan(np.sum(a))
+        # The summary check avoids creating a (potentially huge) array.
+        # However, it will set contains_nan to True for (e.g.) [-inf, 0, +inf].
```
Comment (mdhaber) [inline scipy/stats/stats.py:234]: I do not observe this behavior:
```python3
a = np.random.rand(10)
a[0] = np.nan
print(np.isnan(np.sum(a)))  # True
a[0] = np.inf
print(np.isnan(np.sum(a)))  # False
a[0] = -np.inf
print(np.isnan(np.sum(a)))  # False
a[0] = 0
print(np.isnan(np.sum(a)))  # False
```
If I am missing the point, please clarify.
Also, it seems to me that whether `summary_check` is `True` or `False`, some sort of "summary check" for NaNs is performed. Is there a more descriptive name for this variable?

File: scipy/stats/stats.py [hunk 8]
```diff
@@ -2034,9 +2039,9 @@ def percentileofscore(a, score, kind='rank'):
     Parameters
     ----------
     a : array_like
-        Array of scores to which `score` is compared.
-    score : int or float
-        Score that is compared to the elements in `a`.
+        Array to which `score` is compared.
+    score : array_like
+        Array to compute percentiles for.
```
Comment (mdhaber) [inline scipy/stats/stats.py:2044]:
```suggestion
        Scores to compute percentiles for.
```
I think this emphasizes that this is the 'score' referred to in the name 'percentileofscore'

File: scipy/stats/stats.py [hunk 9]
```diff
@@ -2089,29 +2102,99 @@ def percentileofscore(a, score, kind='rank'):
...
+    # Nan treatment
+    cna, npa = _contains_nan(a, nan_policy, summary_check=False)
+    cns, nps = _contains_nan(score, nan_policy, summary_check=False)
+
+    if (cna or cns) and nan_policy == 'raise':
+        raise ValueError("The input contains nan values")
+
+    elif cns and nan_policy == "omit":
+        raise ValueError("The input scores contains nan values,"
+                         " which is incompatible with the 'omit' policy.")
```
Comment (mdhaber) [inline scipy/stats/stats.py:2142]: I don't think this is consistent with "[A Design Specification for `nan_policy`](https://docs.scipy.org/doc/scipy/reference/dev/api-dev/nan_policy.html)".
image

Is there a reason that NaNs in `score` cannot be removed?

As a minor note, that document consistently uses "`nan`" and "`nan` values". [`np.isnan` documentation](https://numpy.org/doc/stable/reference/generated/numpy.isnan.html) refers to them as "NaN". I know SciPy stats documentation is not consistent about it, but while we're here, I'd suggest "`nan`s" or "NaNs" instead of "nan's".

File: scipy/stats/stats.py [hunk 10]
```diff
@@ -2089,29 +2102,99 @@ def percentileofscore(a, score, kind='rank'):
...
+    # Don't count nans
+    if cna and nan_policy == "omit":
+        a = ma.masked_where(np.isnan(a), a)
+        n = a.count()
+
+    # With `propagate`, an `a` with nans should always yield nan
+    if cna and nan_policy == "propagate":
+        n = 0
+
+    # Cannot compare to empty list ==> nan
```
Comment (mdhaber) [inline scipy/stats/stats.py:2157]: This occurs when the list is not empty but contains NaNs, too.

File: scipy/stats/stats.py [hunk 11]
```diff
@@ -2089,29 +2102,99 @@ def percentileofscore(a, score, kind='rank'):
...
+
+        # Despite masking, Azure pipelines produces warnings
+        # when there are nans (which should not be processed)
+        with suppress_warnings() as sup:
+            sup.filter(RuntimeWarning,
+                       "invalid value encountered in less")
+            sup.filter(RuntimeWarning,
+                       "invalid value encountered in greater")
+
+            # Compute
```
Comment (mdhaber) [inline scipy/stats/stats.py:2174]: I'd suggest being more specific or omitting the comment.

File: scipy/stats/tests/test_stats.py [hunk 12]
```diff
@@ -2753,82 +2753,94 @@ def test_1samp_alternative(self):
...
+
+    @pytest.mark.parametrize("kind, result", [("rank", 40),
+                                              ("mean", 35),
+                                              ("strict", 30),
+                                              ("weak", 40)])
```
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:2765]:
```suggestion
    cases = [("rank", 40), ("mean", 35), ("strict", 30), ("weak", 40)]
    @pytest.mark.parametrize("kind, result", cases)
```
Just a thought.

File: scipy/stats/stats.py [hunk 13]
```diff
@@ -2089,29 +2102,99 @@ def percentileofscore(a, score, kind='rank'):
...
+        n = 0
+
+    # Cannot compare to empty list ==> nan
     if n == 0:
-        return 100.0
-
-    if kind == 'rank':
-        left = np.count_nonzero(a < score)
-        right = np.count_nonzero(a <= score)
-        pct = (right + left + (1 if right > left else 0)) * 50.0/n
```
Comment (mdhaber) [inline scipy/stats/stats.py:2103]: Just to check my understanding of the changes:

- `return` is delayed until the end so NaNs can be added back for implementing `nan_policy='propagate'`
- `np.count_nonzero` is applied along `axis=-1`
- the logic in this line `pct = (right + left + (1 if right > left else 0)) * 50.0/n` is vectorized

The rest - which makes it _appear_ to be a rewrite - is refactoring?

File: scipy/stats/stats.py (hunk 5 above)
Comment (mdhaber) [reply scipy/stats/stats.py]: If we decide that vectorizing this only over `scores` and implementing `nan_policy='omit'` only for `a` is OK, I will try an implementation that avoids masked arrays and see which looks simpler and which is faster, mostly because I'm curious.

**Review by mdhaber — COMMENTED** [review]
Only one substantial comment in the review above, [this one](https://github.com/scipy/scipy/pull/13572/files/785f408afaa00a324c9ef6afa1bd833672d9ff4c#diff-764f0cedd33e98726d12c935495b80333a11e877a5a270d8ce53c9db152dd976).

File: scipy/stats/tests/test_stats.py (hunk 12 above)
Comment (patricknraanes) [reply scipy/stats/tests/test_stats.py:2765]: This does not make so much sense for the other cases (which have "longer" input) so I think I should stick with the current uniformity

File: scipy/stats/stats.py (hunk 7 above)
Comment (patricknraanes) [reply scipy/stats/stats.py:234]: The test cases you run are not detecting the relevant scenario: when the array contains *both* `+inf` and `-inf`, in which case the two methods differ.

I have clarified the comments; in particular reworded "summary" to "summation"

File: scipy/stats/stats.py (hunk 9 above)
Comment (patricknraanes) [reply scipy/stats/stats.py:2142]: > Is there a reason that NaNs in score cannot be removed?

Yes, consider

```
    score = [
        [a, b],
        [c, nan],
    ]

    This should yield:
    result = [
        [pa, pb],
        [pc, ???],
    ]
```

Clearly, the `nan` cannot be omitted. The design specification for `nan_policy` does not seem to consider such cases.

File: scipy/stats/stats.py (hunk 13 above)
Comment (patricknraanes) [reply scipy/stats/stats.py:2103]: Yes, correct on all counts.

File: scipy/stats/stats.py (hunk 10 above)
Comment (patricknraanes) [reply scipy/stats/stats.py:2157]: Yes, and that case is handled by this (copied from the code above):

```
    # Don't count nans
    if cna and nan_policy == "omit":
        a = ma.masked_where(np.isnan(a), a)
        n = a.count()
```

That case is also covered by some of the tests within `test_nans_ok`

**Comment by patricknraanes** [comment]
Apologies for the latency, I was not at my desk(top). I have added more tests specifically for n-D support, and addressed the review by  @mdhaber which required only doc/comment corrections. Some of the remarks by @mdhaber were more major, but IMHO everything was & is in order, though I have left these comments open (not clicked on "resolve conversation" yet).

File: scipy/stats/stats.py (hunk 13 above)
Comment (patricknraanes) [reply scipy/stats/stats.py:1952]: Well, I guess two points was left out:

- `nan_policy` has been implemented.
- More tests

These two points required by far the most work

File: scipy/stats/stats.py (hunk 7 above)
Comment (mdhaber) [reply scipy/stats/stats.py]: I see.

File: scipy/stats/stats.py (hunk 11 above)
Comment (mdhaber) [reply scipy/stats/stats.py:2025]: Please avoid resolving comments that have not been resolved. If you disagree, please reply; otherwise, please consider implementing the suggestion.

File: scipy/stats/tests/test_stats.py (hunk 12 above)
Comment (mdhaber) [reply scipy/stats/tests/test_stats.py:2893]: I suppose in this case I'd break the lines in the variable assignment, e.g.
```python3
cases = [("propagate", [], 1, np.nan),
         ("propagate", [np.nan], 1, np.nan),
         ("propagate", [np.nan], [0, 1, 2], [np.nan, np.nan, np.nan]),
         ("propagate", [1, 2], [1, 2, np.nan], [50, 100, np.nan]),
         ("omit", [1, 2, np.nan], [0, 1, 2], [0, 50, 100]),
         ("omit", [np.nan, np.nan], [0, 1, 2], [np.nan, np.nan, np.nan])])
@pytest.mark.parametrize("policy, a, score, result", cases)
rather than deep in the `parametrize` call, the principle being to break the lines in the simpler statement if possible. But it's a personal preference, so I'll resolve. (Please leave comments open until they're resolved by the reviewer, so that the reviewer can keep track.)

Comment (mdhaber) [reply scipy/stats/tests/test_stats.py:2893]: Actually these later lines are breaking PEP8 standards for line length, so this would be a good solution. Unresolving.

File: scipy/stats/stats.py (hunk 9 above)
Comment (mdhaber) [reply scipy/stats/stats.py:1991]: I see. @tupui you already approved, but what do you think about this?

File: scipy/stats/stats.py (hunk 13 above)
Comment (mdhaber) [reply scipy/stats/stats.py:1952]: I just meant about this part of the code. I asked because the diff makes it look like it was rewritten, but really there are only these changes and the rest is refactoring. I agree that the refactoring makes it look cleaner, so OK to keep it here (but in general I'd recommend refactoring sparingly to speed review).

File: scipy/stats/stats.py (hunk 9 above)
Comment (tupui) [reply scipy/stats/stats.py:1991]: To me it did not make much sense to have nan in the scores as well, but I never work with these so I am not sure. That's why I asked your help here 😅

File: scipy/stats/stats.py (hunk 11 above)
Comment (patricknraanes) [reply scipy/stats/stats.py:2025]: I implemented the suggestion in 178d6fa . Not sure why it's not showing up here (at least by an "obsolete" tag), but I am not very familiar with the Github PR machinery yet.

**Review by patnr — COMMENTED** [review]
I did not mean to start this review (clicked the wrong button). ***Disregard.***

File: scipy/stats/stats.py (hunk 13 above)
Comment (patricknraanes) [reply scipy/stats/stats.py:1952]: Ok, thanks for the pointers 👍

File: scipy/stats/tests/test_stats.py (hunk 12 above)
Comment (patricknraanes) [reply scipy/stats/tests/test_stats.py:2893]: Should be fixed in 080d26c

> Please leave comments open until they're resolved by the reviewer, so that the reviewer can keep track.

Ah ok, I didn't know that was the principle. I thought since GH gives me (who is not a reviewer) the option to "resolve" then I should use it when reaching a decision (with some confidence).

File: scipy/stats/tests/test_stats.py [hunk 14]
```diff
@@ -2881,82 +2881,120 @@ def test_1samp_alternative(self):
...
+    cases = [
+        ("raise", [1, 2, 3, np.nan], [1, 2, 3],
+         "The input contains nan values"),
+        ("raise", [1, 2, 3], [1, 2, 3, np.nan],
+         "The input contains nan values"),
+        ("omit", [1, 2, 3], [1, 2, 3, np.nan],
+         ("The input scores contains nan values,"
+          " which is incompatible with the 'omit' policy.")),
+    ]
+    @pytest.mark.parametrize("policy, a, score, message", )
```
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:2981]:
```suggestion

    @pytest.mark.parametrize("policy, a, score, result", cases)
    def test_nans_ok(self, policy, a, score, result):
        assert_equal(self.f(a, score, nan_policy=policy), result)

    cases = [
        ("raise", [1, 2, 3, np.nan], [1, 2, 3],
         "The input contains nan values"),
        ("raise", [1, 2, 3], [1, 2, 3, np.nan],
         "The input contains nan values"),
        ("omit", [1, 2, 3], [1, 2, 3, np.nan],
         ("The input scores contains nan values,"
          " which is incompatible with the 'omit' policy.")),
    ]

    @pytest.mark.parametrize("policy, a, score, message", )
```
PEP8 wants blank lines here.

File: scipy/stats/tests/test_stats.py (hunk 12 above)
Comment (mdhaber) [reply scipy/stats/tests/test_stats.py:2893]: Some reviewers will mind (like me), but not all. Then again, some reviewers also leave their own comments open indefinitely, even after receiving adequate response. If I agree with the response, I try to resolve them immediately so they don't clutter things. If I don't, I'm glad it's still open so we can discuss it further.

I'll resolve this as the rest of the changes to `test_stats.py` look OK as far as PEP8 is concerned (except for blank lines needed before the functions, which I noted separately).

File: scipy/stats/stats.py (hunk 11 above)
Comment (mdhaber) [reply scipy/stats/stats.py:2025]: It doesn't look like it made it into that commit.

Comment (patricknraanes) [reply scipy/stats/stats.py:2025]: Hmmm... I'm confused. The change is there. See the "files changed" tab, and ctrl-f for "Azure pipelines". I have expanded the comment.

File: scipy/stats/stats.py (hunk 9 above)
Comment (mdhaber) [reply scipy/stats/stats.py:1991]: Yeah... I really don't like `nan_policy = 'omit'`. That's the motivation behind gh-13312; so we fix it once and for all for most functions so we can forget about it. (But that won't work here, so yes, this function does need special consideration.) 

I'm really not sure what to do, but I think when in doubt about interface decisions like this, the mailing list is good thing to try.

File: scipy/stats/stats.py (hunk 11 above)
Comment (mdhaber) [reply scipy/stats/stats.py:2025]: I think I see now. I think you thought I was referring to:
image

This comment was attached to the line:
image

That's the comment I was referring to.

File: scipy/stats/tests/test_stats.py (hunk 14 above)
Comment (patricknraanes) [reply scipy/stats/tests/test_stats.py]: I committed the suggested change here. But will have to wait till tomorrow (when I'm at my desktop) to run flake8

File: scipy/stats/stats.py (hunk 11 above)
Comment (patricknraanes) [reply scipy/stats/stats.py]: Ahhh.... sorry, I thought the lines it was pointing/attached to was a side-effect of other changes I had made. I have made the comment a bit more descriptive in the latest commit. If it's still not quite good, I think I will just remove the comment.

File: scipy/stats/tests/test_stats.py (hunk 14 above)
Comment (patricknraanes) [reply scipy/stats/tests/test_stats.py]: This should be done now.

**Comment by tylerjereddy** [comment]
@mdhaber Is this looking solid now?

**Comment by mdhaber** [comment]
@tylerjereddy I suggested an email to the mailing list [here](https://github.com/scipy/scipy/pull/13572#discussion_r625334934) to get some opinions because this is not consistent with ["A Design Specification for `nan_policy`"](https://docs.scipy.org/doc/scipy/reference/dev/api-dev/nan_policy.html). I don't think that has been done.

**Comment by patricknraanes** [comment]
Am I supposed to email the mailinglist? I'd be happy to, just thought @mdhaber was doing it.

**Comment by mdhaber** [comment]
@tylerjereddy emailed the mailing list, and I talked about it with @WarrenWeckesser, who wrote the policy document, yesterday. He will respond here when he can.

**Comment by tylerjereddy** [comment]
Sounds like this may need some more discussion, so I'll bump the milestone. If I'm wrong in this assessment, you are of course free to clear up the confusion, merge and re-milestone if you get this in before I branch Wednesday.

**Comment by WarrenWeckesser** [comment]
The design decision for how to handle `nan` in the second argument of `percentileofscore` has already been decided in a similar case ([`zmap`](http://scipy.github.io/devdocs/reference/generated/scipy.stats.zmap.html)), and will apply to other functions, so instead of discussing just `percentileofscore` here, I created a separate [issue](https://github.com/scipy/scipy/issues/14221) about it.

File: scipy/stats/stats.py (hunk 9 above)
Comment (mdhaber) [reply scipy/stats/stats.py:1991]: I think gh-14221 says this would be fine as-is. I'd suggest commenting on that issue @patricknraanes to indicate your opinion between the two options. Once we resolve that, i think this is close.

Comment (patricknraanes) [reply scipy/stats/stats.py]: It looks like #14221 will break in favour of "option 1". I therefore made the changes to support it.

Comment (patricknraanes) [reply scipy/stats/stats.py]: Any movement on this?

**Comment by tylerjereddy** [comment]
This has an approval, but also has merge conflicts and appears to be fundamentally blocked by the need for a decision in gh-14221

**Comment by tylerjereddy** [comment]
I will bump the milestone pretty soon if I don't see activity to move that decision forward

**Comment by patricknraanes** [comment]
I'm happy to resolve the merge conflicts once https://github.com/scipy/scipy/issues/14221 is resolved

**Comment by tylerjereddy** [comment]
@patricknraanes I'd say it looks like "option 1" is winning pretty decisively in that discussion now. Not sure if there is time to make the adjustments here to support option 1 and get them approved before branching Monday, but I'll leave the milestone until the last minute here in case you want to try.

File: scipy/stats/stats.py (hunk 9 above)
Comment (tylerjereddy) [reply scipy/stats/stats.py]: I agree that issue seems to have converged to option 1 -- so one of the `stats` folks just needs to give this a final "ok" or are more changes needed to fully support option 1?

**Comment by patricknraanes** [comment]
"Option 1" was fully implemented in https://github.com/scipy/scipy/pull/13572/commits/aa08dcbf8528286563acdd273d950d517d79f1d0

The last 4 commits were to merge with master and improve the documentation in line with "option 1".

**Comment by tylerjereddy** [comment]
@mdhaber @WarrenWeckesser @tupui can you summarize if this is close to merge ready for the `1.9.0` branching later this week?

**Comment by tylerjereddy** [comment]
Ok, the discussion above flip-flops around a lot, but ultimately this PR appears to come out consistent with what our core devs have voted for, and we basically have one core dev in favor here and most comments from Matt seem to have been addressed.

**Comment by tylerjereddy** [comment]
38 commits is a bit heavy here so I will squash-merge I think, thanks--feel free to suggest an appropriate release note entry on the wiki or in my PR over here: https://github.com/scipy/scipy/pull/16248
